drivers/serial/uart_nrfx_uarte.c

```diff
@@ -251,6 +316,475 @@ static int uarte_nrfx_config_get(struct device *dev, struct uart_config *cfg)
+	data->async->rx_buf = buf;
+	nrf_uarte_rx_buffer_set(cfg->uarte_regs, buf, len);
+
+	nrf_uarte_event_clear(cfg->uarte_regs, NRF_UARTE_EVENT_ENDRX);
+	nrf_uarte_event_clear(cfg->uarte_regs, NRF_UARTE_EVENT_RXSTARTED);
+	if (!hw_rx_counting_enabled(data)) {
+		nrf_uarte_event_clear(cfg->uarte_regs, NRF_UARTE_EVENT_RXDRDY);
+	}
+
+	nrf_uarte_task_trigger(cfg->uarte_regs, NRF_UARTE_TASK_STARTRX);
+	data->async->rx_enabled = true;
+	return 0;
+}
+
+#define UARTE_ERROR_FROM_MASK(mask) \
+	(mask & NRF_UARTE_ERROR_OVERRUN_MASK ? UART_ERROR_OVERRUN	\
+	 : mask & NRF_UARTE_ERROR_PARITY_MASK ? UART_ERROR_PARITY	\
+	 : mask & NRF_UARTE_ERROR_FRAMING_MASK ? UART_ERROR_FRAMING	\
+	 : mask & NRF_UARTE_ERROR_BREAK_MASK ? UART_BREAK		\
+	 : 0)
+
+static void error_isr(struct device *dev)
+{
+	const struct uarte_nrfx_config *cfg = get_dev_config(dev);
+	u32_t err = nrf_uarte_errorsrc_get_and_clear(cfg->uarte_regs);
+	struct uart_event evt = {
+		.type = UART_RX_STOPPED,
+		.data.rx_stop.reason = UARTE_ERROR_FROM_MASK(err),
+	};
+	user_callback(dev, &evt);
+}
+
+static void rxstarted_isr(struct device *dev)
+{
+	struct uarte_nrfx_data *data = get_dev_data(dev);
+	struct uart_event evt = {
+		.type = UART_RX_BUF_REQUEST,
+	};
+	user_callback(dev, &evt);
+	if (data->async->rx_timeout) {
+		data->async->rx_timeout_left = data->async->rx_timeout;
+		k_delayed_work_submit(&data->async->rx_timeout_work,
+				data->async->rx_timeout_slab);
+	}
+}
+
+static void endrx_isr(struct device *dev)
+{
+	struct uarte_nrfx_data *data = get_dev_data(dev);
+	const struct uarte_nrfx_config *cfg = get_dev_config(dev);
+
+	if (!data->async->rx_enabled) {
+		return;
+	}
+	if (data->async->rx_next_buf) {
+		nrf_uarte_task_trigger(cfg->uarte_regs, NRF_UARTE_TASK_STARTRX);
+	}
+	k_delayed_work_cancel(&data->async->rx_timeout_work);
+	size_t rx_len = nrf_uarte_rx_amount_get(cfg->uarte_regs);
+	struct uart_event evt = {
+		.type = UART_RX_RDY,
+		.data.rx.buf = data->async->rx_buf,
+		.data.rx.len = rx_len - data->async->rx_offset,
+		.data.rx.offset = data->async->rx_offset,
+	};
+	user_callback(dev, &evt);
+	data->async->rx_total_byte_cnt += evt.data.rx.len;
+	evt.type = UART_RX_BUF_RELEASED;
+	evt.data.rx_buf.buf = data->async->rx_buf;
+	user_callback(dev, &evt);
+	if (data->async->rx_next_buf) {
+		data->async->rx_buf = data->async->rx_next_buf;
+		data->async->rx_size = data->async->rx_next_size;
+	} else {
+		data->async->rx_buf = 0;
+		data->async->rx_size = 0;
+		if (hw_rx_counting_enabled(data)) {
+			gppi_channel_disable(data->async->rx_cnt.ppi);
+			nrfx_timer_clear(&data->async->timer);
+		} else {
+			data->async->rx_cnt.cnt = 0;
+		}
+		data->async->rx_total_byte_cnt = 0;
+		evt.type = UART_RX_DISABLED;
+		user_callback(dev, &evt);
+	}
+	data->async->rx_total_user_byte_cnt = data->async->rx_total_byte_cnt;
+	data->async->rx_next_size = 0;
+	data->async->rx_next_buf = NULL;
+	data->async->rx_offset = 0;
+}
+
+static void rxto_isr(struct device *dev)
+{
```

Comment:
It would be good to add a comment that this handler is called when the reception is interrupted, in contrary to finishing the reception after filling all provided buffers, in which case the events `UART_RX_BUF_RELEASED` and `UART_RX_DISABLED` are reported from `endrx_isr`. It's not that obvious.